### PR TITLE
Add higher order autodiff for getindex

### PR DESCRIPTION
**Pull request by Keno**
This fixed higher order autodiff or getindex (and by extension vcat, since
it's backward pass uses getindex). This makes tracker able to handle
the third order derivatives needed to implement [1].

[1] Physics-Informed Generative Adversarial Networks for Stochastic Differential Equations. https://arxiv.org/abs/1811.02033

### CODE REVIEW

**Comment by AzamatB** [comment]
@Keno are you implementing [1]? Will you add it to the model zoo?

**Comment by Keno** [comment]
We're working on an extension of that work. I can see if we can put the baseline implementation in the model zoo.

**Comment by AzamatB** [comment]
@Keno thanks a lot! Would be great to have that implementation in the model zoo :)

File: src/tracker/lib/array.jl [hunk 1]
```diff
@@ -95,14 +95,27 @@ end
 
 Base.getindex(xs::TrackedArray, i...) = track(getindex, xs, i...)
 
-@grad function getindex(xs::AbstractArray, i...)
-  data(xs)[i...], function (Δ)
-    Δ′ = zero(xs)
-    Δ′[i...] = data(Δ)
-    (nobacksies(:getindex, Δ′), map(_->nothing, i)...)
+struct getindex_back{T,S}
```
Comment (MikeInnes) [inline src/tracker/lib/array.jl]: Would prefer this to follow the usual naming conventions.

File: src/tracker/lib/array.jl [hunk 2]
```diff
@@ -95,14 +95,27 @@ end
...
+    xs::T
+    i::S
+end
+
+function (g::getindex_back)(Δ)
+  Δ′ = zero(g.xs)
+  Δ′[g.i...] = data(Δ)
```
Comment (MikeInnes) [inline src/tracker/lib/array.jl]: Calling `data` here drops the gradient for `Δ`, which you won't want here.

File: test/tracker.jl [hunk 3]
```diff
@@ -36,6 +36,15 @@ gradtest(f, dims...) = gradtest(f, rand.(Float64, dims)...)
 
 @testset "indexing & slicing" begin
   gradtest(x->view(x, 1:2, 1:2), rand(4, 4))
+
+  # Nested AD for getindex
+  @test gradient([1.0, 2.0, 3.0]; nest=true) do x
```
Comment (MikeInnes) [inline test/tracker.jl]: Would be good to have a numerical check (or one against ForwardDiff, if that's too unstable).

File: src/tracker/lib/array.jl [hunk 4]
```diff
@@ -95,14 +95,27 @@ end
...
+  (Δ′, map(_->nothing, g.i)...)
+end
+
+@grad function (g::getindex_back)(Δ)
```
Comment (MikeInnes) [inline src/tracker/lib/array.jl]: This doesn't actually get called at the moment.

File: src/tracker/lib/array.jl [hunk 5]
```diff
@@ -95,14 +95,27 @@ end
...
+  g(Δ), function(Δ′′)
+    (Δ′′[g.i...], map(_->nothing, g.i)...)
```
Comment (MikeInnes) [inline src/tracker/lib/array.jl]: The `nothing`s aren't necessary here.

File: test/tracker.jl (hunk 3 above)
Comment (Keno) [reply test/tracker.jl]: yeah, I realized I got it wrong ;). Will push a fix shortly.

**Comment by Keno** [comment]
> Thanks, this seems like generally the right idea. If your model is training with this though, it shouldn't be :)

It's not. I had also realized I got it wrong. Thanks for the review though. I'll fix it up.

**Comment by Keno** [comment]
Alright, try this.

**Comment by Drvi** [comment]
Since the `∇getindex` struct holds the gradient and the destination indices, I think MR is not that far from the idea in #589 (I try to avoid allocating the full array of zeros and then copying the grad into it by specializing the `accum!()` function to just add the grad inplace). It really helped performance in my case.

The logic for getindex is getting more complex here, so I'll see if I can make #589 compatible.

File: src/tracker/lib/array.jl [hunk 6]
```diff
@@ -95,12 +95,27 @@ end
...
+
+@grad function (g::∇getindex)(Δ)
+  z, back = g(data(Δ)), function(Δ′′)
+    (Δ′′[1][g.i...],)
   end
+  z, back
+end
+
+@grad function getindex(xs::AbstractArray, i...)
+  data(xs)[i...], ∇getindex(xs, i)
```
Comment (MikeInnes) [inline src/tracker/lib/array.jl]: You could simplify this by having `∇getindex` return an array and wrapping it to return the right tuple. Then the changes to TrackedTuples and backprop wouldn't be needed.

Comment (Keno) [reply src/tracker/lib/array.jl]: so have a closure here that returns the tuple or sth to that extent?

Comment (MikeInnes) [reply src/tracker/lib/array.jl]: Yeah exactly

**Pull request closed without merging**